Fix zoom ratio calculation

ulteq opened the pull request:
This simplifies the logic behind the automatic zoom ratio calculation, which is used by both `--auto-zoom` and `--scale-down`.

When merged this will:
* Improve the performance (especially in fullscreen mode)
* Fix the `w` command when both `--scale-down` and `--keep-zoom-vp` are enabled
* Fix `--auto-zoom` not being triggered on window resize events when `--scale-down` is enabled
* Fix `--auto-zoom` not being applied to the first image
* Fix `--auto-zoom` conflicting with manual zoom
* Fix `--zoom <percent> --auto-zoom` logic according to the `man feh` description
* Fix `feh_draw_checks` not taking the zoom level into account properly
* Fix `feh_draw_checks` being called before the zoom level is adjusted
* Prevent `--zoom <percent>` from blocking `--scale-down` in fullscreen / fixed geometry mode
* Prevent `--keep-zoom-vp` from blocking the dynamic window resizing mechanism
* Prevent automatic recalculation of the zoom ratio when `--keep_zoom_vp` is enabled
* Prevent `winwid-> im_x` and `winwid-> im_y` from being set to invalid values
(when switching between images of different sizes)
* Simplify the way `--keep-zoom-vp` is handled internally

Fixes: #229
Fixes: #244
Fixes: #307 

Closes: https://github.com/derf/feh/pull/358
Closes: https://github.com/derf/feh/pull/361
Closes: https://github.com/derf/feh/pull/362

derf commented:
Hmm. I'm happy that someone is attempting a cleanup of this mess, though this still needs some testing (and, possibly, polishing).

Right now the first slide of a `feh --fullscreen` slideshow is not centered properly. I'll let you know if I find anything else (or come up with a patch for it, in which case I'll happily merge the request unless I find other issues).

Expect delays though, I don't have much time at the moment.

ulteq commented:
> though this still needs some testing 

I absolutely agree.

> Right now the first slide of a feh --fullscreen slideshow is not centered properly.

I will look into it.

> Expect delays though

No problem. We don't have to rush this.

ulteq commented:
I backported those fixes from this commit: 4571b34c9461ba9bb6309983e5be79570c3e6f9e

ulteq commented:
@derf This is ready for a second review.

derf commented:
hmm, this still causes some behaviour changes in feh. fullscreen mode no longer zooms out images which are too large, and due to the re-use of the geometry flags settings leaving fullscreen mode will cause non-tiling feh windows to be as big as the screen (instead of fitted to the size of the image, which is usually smaller). Also, in non-tiling mode, feh will no longer resize its window when changing images.

I've yet to decide which of these I consider to be an issue and which I don't (and do more testing), so this is just the current status to make sure i don't forget it.

ulteq commented:
> fullscreen mode no longer zooms out images which are too large

That was easy to fix.

> Also, in non-tiling mode, feh will no longer resize its window when changing images.

This happens because I enabled fixed geometry mode by default, to avoid the flickering in tiling windows:
https://github.com/derf/feh/pull/364/commits/203e8fff263496996cad6b5cffc2cebe614ac190#diff-24777376c7590882b07457017fd284ceL326

I think we should revert that and then either try to implement an automatic detection routine to differentiate between floating and non floating windows, or allow the user to enable fixed geometry mode by default via an option as suggest by @0xmatthias.

ulteq commented:
> fullscreen mode no longer zooms out images which are too large

I modified this condition in order to replicate the old behavior more closely: https://github.com/derf/feh/pull/364/commits/203e8fff263496996cad6b5cffc2cebe614ac190#diff-24777376c7590882b07457017fd284ceR456

> Also, in non-tiling mode, feh will no longer resize its window when changing images.
> leaving fullscreen mode will cause non-tiling feh windows to be as big as the screen

Both of this happened because I enabled fixed geometry mode by default in order to avoid the flickering in tiling windows: https://github.com/derf/feh/pull/364/commits/203e8fff263496996cad6b5cffc2cebe614ac190#diff-24777376c7590882b07457017fd284ceL326

I have reverted it and will try to find another solution to fix the flickering issue with tiling windows.

ulteq commented:
Feel free to drop this commit: e7050e08384858054203d4248dae2b1e59d29e8f, but I think it would be a great addition for those who use feh to read comics.

ulteq commented:
Please let me know if there is anything I can do to help you make this pull request ready for merge.

derf commented:
Actually, just poking me again was sufficient. I couldn't find any issues with this version :)

Thanks a lot for the patches and sorry for the long wait!
